components/hybrid_engine/src/observer.rs

```diff
@@ -0,0 +1,259 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::{Arc, Mutex};
+
+use engine_traits::{CacheRange, KvEngine, RangeCacheEngineExt};
+use kvproto::{metapb::Region, raft_cmdpb::AdminCmdType, raft_serverpb::RaftApplyState};
+use raft::StateRole;
+use raftstore::coprocessor::{
+    AdminObserver, ApplyCtxInfo, ApplySnapshotObserver, BoxAdminObserver, BoxApplySnapshotObserver,
+    BoxCmdObserver, BoxQueryObserver, BoxRoleObserver, Cmd, CmdBatch, CmdObserver, Coprocessor,
+    CoprocessorHost, ObserveLevel, ObserverContext, QueryObserver, RegionState, RoleObserver,
+};
+
+#[derive(Clone)]
+pub struct Observer {
+    pending_evict: Arc<Mutex<Vec<CacheRange>>>,
+    cache_engine: Arc<dyn RangeCacheEngineExt + Send + Sync>,
+}
+
+impl Observer {
+    pub fn new(cache_engine: Arc<dyn RangeCacheEngineExt + Send + Sync>) -> Self {
+        Observer {
+            pending_evict: Arc::default(),
+            cache_engine,
+        }
+    }
+
+    pub fn register_to(&self, coprocessor_host: &mut CoprocessorHost<impl KvEngine>) {
+        // This observer does not need high priority, use the default 100.
+        let priority = 100;
+        coprocessor_host
+            .registry
+            .register_cmd_observer(priority, BoxCmdObserver::new(self.clone()));
+        // Evict cache when a peer applies ingest sst.
+        coprocessor_host
+            .registry
+            .register_query_observer(priority, BoxQueryObserver::new(self.clone()));
+        // Evict cache when a peer splits and merges.
+        coprocessor_host
+            .registry
+            .register_admin_observer(priority, BoxAdminObserver::new(self.clone()));
+        // Evict cache when a peer applies snapshot.
+        coprocessor_host.registry.register_apply_snapshot_observer(
+            priority,
+            BoxApplySnapshotObserver::new(self.clone()),
+        );
+        // Evict cache when a leader steps down.
+        coprocessor_host
+            .registry
+            .register_role_observer(priority, BoxRoleObserver::new(self.clone()));
+    }
+
+    fn post_exec_cmd(
+        &self,
+        ctx: &mut ObserverContext<'_>,
+        cmd: &Cmd,
+        state: &RegionState,
+        apply: &mut ApplyCtxInfo<'_>,
+    ) {
+        if !self.cache_engine.range_cache_engine_enabled() {
+            return;
+        }
+        // Evict caches for successfully executed ingest commands and admin
+        // commands that change region range.
+        if apply.pending_handle_ssts.is_some()
+            || (state.modified_region.is_some()
+                && matches!(
+                    cmd.request.get_admin_request().get_cmd_type(),
+                    AdminCmdType::BatchSplit
```

Comment:
The range can still servce requests after batch split (it still has complete data view), so it should not evict.